bulk_indexer_pool_test.go

```diff
@@ -0,0 +1,279 @@
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package docappender
+
+import (
+	"context"
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+	"testing"
+	"time"
+
+	"github.com/elastic/go-docappender/v2/docappendertest"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+)
+
+func TestBulkIndexerPoolConcurrent(t *testing.T) {
+	// This test ensures that the pool can be used concurrently
+	// without blocking on Get() under load.
+	// Additionally, it ensures that the pool can continue servicing
+	// up to the minimum number of indexers per ID even if the maximum
+	// number of "leased" indexers is reached.
+	// This ensures Quality of Service for the minimum number of indexers
+	// per ID.
+	type testCfg struct {
+		latency time.Duration // Latency between each get
+		draw    int           // The number of pool.Get() calls.
+	}
+	var mu sync.Mutex
+	uniques := map[string]struct{}{}
+	cfg := BulkIndexerConfig{
+		Client: docappendertest.NewMockElasticsearchClient(t, func(http.ResponseWriter, *http.Request) {
+			mu.Lock()
+			uniques["invalid"] = struct{}{}
+			mu.Unlock()
+		}),
+	}
+	require.NoError(t, cfg.Validate())
+	test := func(t testing.TB, guaranteed, localMax, total int) {
+		pool := NewBulkIndexerPool(guaranteed, localMax, total, cfg)
+		parameters := map[string]testCfg{
+			// Fast indexer up to the maximum (total) number of indexers
+			// allowed in the pool.
+			"fast": {latency: time.Nanosecond, draw: localMax},
+			// Slow will pull up to the minimum number of indexers.
+			"slow": {latency: time.Millisecond / 2, draw: guaranteed},
+			// Slower will pull up to the minimum number of indexers.
+			"slower": {latency: 5 * time.Millisecond, draw: guaranteed},
+		}
+		var wg sync.WaitGroup
+		for id, p := range parameters {
+			wg.Add(1)
+			go func(id string, p testCfg) {
+				defer wg.Done()
+				pool.Register(id)
+				for i := 0; i < p.draw; i++ {
+					t.Log("start", id, pool.count(id))
+					indexer, _ := pool.Get(context.Background(), id)
+					indexer.ResetClient(docappendertest.NewMockElasticsearchClient(t, func(http.ResponseWriter, *http.Request) {
+						mu.Lock()
+						defer mu.Unlock()
+						uniques[fmt.Sprint(id)] = struct{}{}
+					}))
+					require.NotNil(t, indexer)
+					indexer.Add(BulkIndexerItem{
+						Index: "test",
+						Body:  strings.NewReader(`{"foo":"bar"}`),
+					})
+					indexer.Flush(context.Background())
+					time.Sleep(p.latency)
+					defer func(indexer *BulkIndexer) {
+						pool.Put(id, indexer)
+						t.Log("put", id, pool.count(id))
+					}(indexer)
+				}
+				t.Log("end", id, pool.count(id))
+			}(id, p)
+		}
+		wg.Wait()
+		// Now we deregister all IDs and ensure that the pool is empty.
+		for id := range parameters {
+			c := pool.Deregister(id)
+			require.Equal(t, len(c), 0, id)
+			assert.Nil(t, <-c) // Assert nil (closed).
```

Comment:
I assume the channel will never have nils in it on normal circumstances, hence why we can check nil for closed? If so, ignore suggestion below:

```suggestion
            _, open := <-c
			assert.False(t, open) // Assert closed channel.
```

Reply:
Closing a channel returns the zero value of the channel type. In this case we have a `chan *BulkIndexer`, so when the channel is closed, it returns the zero value of `*BulkIndexer`, which is `nil`.

Reply:
Yes, I meant that if we were to do:
```
c <- nil
assert.Nil(t, <-c) // Also passes
```

But I think in this case we basically never pass nil into channel, so either way works.